# PopPUNK-mod
PopPUNK-mod is used for fitting models of pangenome dynamics to PopPUNK pairwise core vs. accessory distance distributions.

PopPUNK-mod uses Approximate Bayesian Computation (ABC) using [BOLFI](https://jmlr.org/papers/v17/15-017.html), part of the [ELFI](https://www.jmlr.org/papers/volume19/17-374/17-374.pdf) package

## Installation

First install all dependencies. This is easiest using [micromamba](https://mamba.readthedocs.io/en/latest/installation/micromamba-installation.html).

```
micromamba env create -f environment.yml
micromamba activate pp_mod
```

This will create and activate the pp_mod environment. Then, clone this repository:

```
git clone https://github.com/samhorsfield96/PopPUNK-mod.git && cd PopPUNK-mod
```

Finally, install [Pansim](https://github.com/bacpop/Pansim), a fast pangenome dynamics simulator used by PopPUNK-mod. Installation instructions can be found on the Pansim G[Github](https://github.com/bacpop/Pansim).

## Running Pansim

You can run a wrapper around Pansim using `pansim.py`, ensuring you point to the Pansim executable:

```
python pansim.py --pansim_exe /path/to/pansim --outpref simulations
```

### Outputs

This will generate:
- All outputs generated by Pansim (see [repository](https://github.com/bacpop/Pansim))
- `<outpref>_sim.png`; a graph showing the distances, along with a asymototic curve fit

### Command-line options

```
usage: python pansim.py [-h] [--core_size CORE_SIZE] [--pan_genes PAN_GENES] [--core_genes CORE_GENES] [--core_mu CORE_MU] [--rate_genes1 RATE_GENES1]     [--rate_genes2 RATE_GENES2]
                                    [--prop_genes2 PROP_GENES2] [--prop_positive PROP_POSITIVE] [--pos_lambda POS_LAMBDA] [--neg_lambda NEG_LAMBDA] [--pop_size POP_SIZE] [--n_gen N_GEN]
                                    [--avg_gene_freq AVG_GENE_FREQ] [--HR_rate HR_RATE] [--HGT_rate HGT_RATE] [--competition_strength COMPETITION_STRENGTH] [--max_distances MAX_DISTANCES]
                                    [--outpref OUTPREF] [--seed SEED] [--threads THREADS] --pansim_exe PANSIM_EXE

Wrapper around Pansim.

options:
  -h, --help            show this help message and exit

Input/Output options:
  --core_size CORE_SIZE
                        Number of positions in core genome. Default = 1200000
  --pan_genes PAN_GENES
                        Number of genes in pangenome, including core and accessory genes. Default = 6000
  --core_genes CORE_GENES
                        Number of core genes in pangenome only. Default = 2000
  --core_mu CORE_MU     Maximum pairwise distance for core genome. Default = 0.05
  --rate_genes1 RATE_GENES1
                        Average number of accessory pangenome that mutates per generation in gene compartment 1. Must be >= 0.0. Default = 10.0
  --rate_genes2 RATE_GENES2
                        Average number of accessory pangenome that mutates per generation in gene compartment 2. Must be >= 0.0. Default = 10.0
  --prop_genes2 PROP_GENES2
                        Proportion of pangenome made up of compartment 2 genes. Must be 0.0 <= X <= 0.5. Default = 2.0
  --prop_positive PROP_POSITIVE
                        Proportion of pangenome made up of compartment 2 genes. Must be 0.0 <= X <= 0.5. Default = 2.0. If negative, neutral selection is simulated.
  --pos_lambda POS_LAMBDA
                        Lambda value for exponential distribution of positively selected genes. Must be > 0.0
  --neg_lambda NEG_LAMBDA
                        Lambda value for exponential distribution of negatively selected genes. Must be > 0.0
  --pop_size POP_SIZE   Population size for Wright-Fisher model. Default = 1000
  --n_gen N_GEN         Number of generations for Wright-Fisher model. Default = 100
  --avg_gene_freq AVG_GENE_FREQ
                        Average gene frequency in accessory genome. Default = "0.5"
  --HR_rate HR_RATE     Homologous recombination rate, as number of core sites transferred per core genome mutation.Default=0.0
  --HGT_rate HGT_RATE   HGT rate, as number of accessory sites transferred per core genome mutation.Default=0.0
  --competition_strength COMPETITION_STRENGTH
                        Run simulator with competition.
  --max_distances MAX_DISTANCES
                        Number of distances to sample with Pansim. Default = 100000
  --outpref OUTPREF     Output prefix. Default = "sim"
  --seed SEED           Seed for random number generation. Default = 254.
  --threads THREADS     Number of threads. Default = 1
  --pansim_exe PANSIM_EXE
                        Path to pansim executable.
```

## Running PopPUNK-mod

PopPUNK-mod requires a target distribution to fit to, as well as at least one parameter to fit. All other parameters can be fixed.

```
python poppunk_mod.py --pansim_exe /path/to/pansim --outpref predictions
```